src/algorithms/time_evolution/fullenv_truncation.jl

```diff
@@ -0,0 +1,251 @@
+```
+Then (each index corresponds to a pair of fused indices)
+```
+    F(r,r†) = |p† r|² / (r† B r)
+            = (r† p) (p† r) / (r† B r)
+```
+which is maximized when
+```
+    ∂F/∂r̄ * (r† B r)²
+    = p (p† r) (r† B r) - |p† r|² (B r) = 0
+```
+Note that `B` is positive (consequently `B† = B`). 
+Then the solution for the vector `r` is
+```
+    r = B⁻¹ p
+```
+We can verify that (using `B† = B`)
+```
+    ∂F/∂r̄ * (r† B r)²
+    = p (p† B⁻¹ p) (p† B⁻¹ B B⁻¹ p) - |p† B⁻¹ p|² (B B⁻¹ p) 
+    = 0
+```
+Then the bond matrix `u s v†` is updated by truncated SVD:
+```
+    - u ← r -    ==>    - u ← s ← v† -
+```
+
+## Optimization of `l`
+
+The process is entirely similar. 
+Define the vector `p` and the positive map `B` as
+```
+        ┌---------------┐   ┌-----------------------┐
+        |   ┌---┐       |   |   ┌----┐              |
+        └---|   |-------┘   └---|    |------b0† ----┘
+            | p†|         =     |benv|
+        ┌---|   |-  ←-o-┐   ┌---|    |--    ←-o- v†-┐
+        |   └---┘       |   |   └----┘              |
+        └---------------┘   └-----------------------┘
+
+        ┌---------------┐   ┌-----------------------┐
+        |   ┌---┐       |   |   ┌----┐              |
+        └---|   |-  →-o-┘   └---|    |--    →-o- v -┘
+            | B |         =     |benv|
+        ┌---|   |-  ←-o-┐   ┌---|    |--    ←-o- v†-┐
+        |   └---┘       |   |   └----┘              |
+        └---------------┘   └-----------------------┘
+```
+Here `o` is the parity tensor (twist) necessary for fermions. 
+Then (each index corresponds to a pair of fused indices)
+```
+    F(l,l†) = |p† l|² / (l† B l)
+```
+which is maximized when
+```
+    l = B⁻¹ p
+```
+Then the bond matrix `u s v†` is updated by SVD:
+```
+    - l ← v† -   ==>    - u ← s ← v† -
+```
+
+## Returns
+
+The SVD result of the new bond matrix `u`, `s`, `vh`.
+"""
+function fullenv_truncate(
+    b0::AbstractTensorMap{T,S,1,1}, benv::BondEnv{T,S}, alg::FullEnvTruncation
+) where {T<:Number,S<:ElementarySpace}
+    verbose = (alg.check_interval > 0)
+    # `benv` is assumed to be positive; here we only check codomain(benv) == domain(benv).
+    @assert codomain(benv) == domain(benv)
+    time00 = time()
+    # initialize u, s, vh with truncated SVD
+    u, s, vh = tsvd(b0; trunc=alg.trscheme)
+    # normalize `s` (bond matrices can always be normalized)
+    s /= norm(s, Inf)
+    s0 = deepcopy(s)
+    Δfid, Δs, fid, fid0 = NaN, NaN, 0.0, 0.0
+    for iter in 1:(alg.maxiter)
+        time0 = time()
+        # update `← r -  =  ← s ← v† -`
+        @tensor r[-1 -2] := s[-1 1] * vh[1 -2]
+        @tensor p[-1 -2] := conj(u[1 -1]) * benv[1 -2; 3 4] * b0[3 4]
+        @tensor B[-1 -2; -3 -4] := conj(u[1 -1]) * benv[1 -2; 3 -4] * u[3 -3]
+        _linearmap_twist!(p)
+        _linearmap_twist!(B)
+        r, info_r = linsolve(x -> B * x, p, r, 0, 1)
+        @tensor b1[-1; -2] := u[-1 1] * r[1 -2]
+        u, s, vh = tsvd(b1; trunc=alg.trscheme)
```

Comment:
```suggestion
        u, s, vh = tsvd!(b1; trunc=alg.trscheme)
```
I think you can safely destroy `b1` here and avoid a copy?

Reply:
In the following I need to calculate the fidelity between the initial `b0` and the optimized `b1`, so I can't destroy it here.